File: fsharp/triangle/Triangle.fs
module Triangle

    let private (|IsTriangle|_|) (triangle:double list) =
        match triangle with
        | [a:double; b:double; c:double] when a + b > c && b + c > a && a + c > b ->
            [a; b; c]
            |> Some
        | _ -> None

    let equilateral (triangle:double list):bool =
        match triangle with
        | IsTriangle [a:double; b:double; c:double] -> a = b && b = c
        | _ -> false

    let isosceles (triangle:double list):bool =
        match triangle with
        | IsTriangle [a:double; b:double; c:double] -> a = b || b = c || c = a
        | _ -> false

    let scalene (triangle:double list):bool =
        match triangle with
        | IsTriangle [a:double; b:double; c:double] -> a <> b && b <> c && c <> a
        | _ -> false

File: solutions/scala/triangle/1/src/main/scala/Triangle.scala
class Triangle(a:Double, b:Double, c:Double)
{
    private def isTriangle:Boolean = this.a + this.b > this.c && this.b + this.c > this.a && this.a + this.c > this.b

    def equilateral:Boolean = this.isTriangle && this.a == this.b && this.b == this.c

    def isosceles:Boolean = this.isTriangle && this.a == this.b || this.b == this.c || this.c == this.a

    def scalene:Boolean = this.isTriangle && this.a != this.b && this.b != this.c && this.c != this.a
}